tck/src/main/java/org/eclipse/microprofile/config/tck/CDIPlainInjectionTest.java

```diff
@@ -0,0 +1,235 @@
+        assertThat(bean.longProperty, is(equalTo(10L)));
+        assertThat(bean.floatProperty, is(floatCloseTo(10.5f, 0.1f)));
+        assertThat(bean.doubleProperty, is(closeTo(11.5, 0.1)));
+        assertThat(bean.dateProperty, is(equalTo(CDITestsFunctions.toDate("2017-01-30"))));
+        assertThat(bean.localDateProperty, is(CDITestsFunctions.toDate("2016-01-30")));
+        assertThat(bean.dateTimeProperty, is(CDITestsFunctions.toDate("2015-01-30T10:00")));
+    }
+
+    @Test
+    public void can_inject_optional_values() {
+        ensure_property_undefined("non-existent.string.property");
+        ensure_property_defined("my.string.property", "text");
+
+        OptionalValuesBean bean = get_bean_of_type(OptionalValuesBean.class);
+
+        assertThat(bean.noStringProperty.isPresent(), is(false));
+        assertThat(bean.stringProperty.isPresent(), is(true));
+        assertThat(bean.stringProperty.get(), is(equalTo("text")));
+    }
+        
+    @Test
+    public void can_inject_dynamic_values_via_CDI_provider() {
+        clear_all_property_values();
+
+        DynamicValuesBean bean = get_bean_of_type(DynamicValuesBean.class);
+
+        assertThat(bean.getIntProperty(), is(nullValue()));
+
+        ensure_all_property_values_are_defined();
+
+        assertThat(bean.getIntProperty(), is(equalTo(5)));
+    }
+
+    @Test
+    public void can_inject_dynamic_values_via_custom_interface() {
+        clear_all_property_values();
+
+        CustomConfigBean bean = get_bean_of_type(CustomConfigBean.class);
+
+        assertThat(bean.getIntProperty(), is(nullValue()));
+        assertThat(bean.getStringProperty(), is(nullValue()));
+
+        ensure_all_property_values_are_defined();
+
+        assertThat(bean.getIntProperty(), is(equalTo(5)));
+        assertThat(bean.getStringProperty(), is(equalTo("text")));
+    }
+
+    private void ensure_all_property_values_are_defined() {
+        ensure_property_defined("my.string.property", "text");
+        ensure_property_defined("my.boolean.property", "true");
+        ensure_property_defined("my.int.property", "5");
+        ensure_property_defined("my.long.property", "10");
+        ensure_property_defined("my.float.property", "10.5");
+        ensure_property_defined("my.double.property", "11.5");
+        ensure_property_defined("my.date.property", "2017-01-30");
+        ensure_property_defined("my.localdate.property", "2016-01-30");
+        ensure_property_defined("my.datetime.property", "2015-01-30T10:00");
+    }
+
+    private void ensure_property_defined(String key, String value) {
+        // setting configuration via system properties
+        System.setProperty(key, value);
+    }
+
+    private void ensure_property_undefined(String key) {
+        // clearing configuration in system properties if previously set
+        System.getProperties().remove(key);
+    }
+    
+    private static <T> T get_bean_of_type(Class<T> beanClass) {
```

Comment:
I know I can use inject in Arquilian test, but I wanted to get the implementation details out of the way and write a clean and legible test suite. Tha's also why I moved methods like this to a separate util class.